src/main/java/io/usethesource/vallang/impl/persistent/PersistentHashIndexedBinaryRelation.java

```diff
@@ -644,18 +644,23 @@ else if (values instanceof Set) {
       else {
         throw new IllegalArgumentException("Unexpected map entry");
       }
-      // we mark ourselves as done before we did it, 
-      // so we don't do the <a,a> that causes an extra round
-      done.add(lhs); 
       IValue rhs;
       while ((rhs = todo.poll()) != null) {
+        if (rhs.equals(lhs)) {
```

Comment:
this could possibly be an expensive check. Why not try `lhs == rhs`, which would at least eliminate all the self-loops we generate ourselves, but be a lot cheaper.

Reply:
Agreed, that was something I was wondering too. but there aren't that many guarantees about it being the same instance, sure it might cover some cases, but would miss others. It's just that a self edge causes quite some extra useless inserts. Well, maybe the object equality is good enough.

Reply:
I'm more worried about the .equals could have introduced a new bottleneck in some cases where the values are big and nested, and not shared. So let's keep it at `==` indeed, and we get the benefit of your fast path without the risk.